File: ffi/python/blsct/ctx_id.py
from . import blsct
from .managed_obj import ManagedObj
from .serializable import Serializable
from typing import Any, override, Type, Self

class CtxId(ManagedObj, Serializable):
  """
  Represents the transaction ID of a CMutableTransaction

  >>> from blsct import CTxId, TX_ID_SIZE
  >>> import secrets
  >>> hex = secrets.token_hex(TX_ID_SIZE)
  >>> tx_id = CTxId.deserialize(hex)
  >>> tx_id.to_hex()
  'f60b407e98916361594ecd53d7bed716fb901570815c323244da8d4189833df3'  # doctest: +SKIP
  """
  @override
  def serialize(self) -> str:
    """Serialize the CtxId object to a hexadecimal string."""
    buf = blsct.cast_to_uint8_t_ptr(self.value())
    return blsct.to_hex(buf, blsct.TX_ID_SIZE)

  @classmethod
  @override
  def deserialize(
    cls: Type[Self],
    hex: str,
  ) -> Self:
    """Create a TxId from a hexadecimal string."""
    if len(hex) != blsct.TX_ID_SIZE * 2:
      raise ValueError(f"Invlid TxId hex length. Expected {blsct.TX_ID_SIZE * 2}, but got {len(hex)}.")
    obj = blsct.hex_to_malloced_buf(hex) 
    return cls(obj)

  @override
  def value(self):
    return blsct.cast_to_uint8_t_ptr(self.obj)

  @classmethod
  @override
  def default_obj(cls: Type[Self]) -> Any:
    raise NotImplementedError("Cannot create a CtxId without required parameters.")

File: ffi/ts/src/ctxId.ts
import {
  castToUint8_tPtr,
  CTX_ID_SIZE,
  hexToMallocedBuf,
  toHex,
} from './blsct'

import { ManagedObj } from './managedObj'

/** Represents the transaction ID of a CMutableTransaction
 *
 * Examples:
 * ```ts
 * const { CTxId, CTX_ID_SIZE } = require('navio-blsct')
 * const { randomBytes } = require('crypto')
 * const cTxIdHex = randomBytes(CTX_ID_SIZE).toString('hex')
 * const cTxId = CTxId.deserialize(cTxIdHex)
 * const ser = cTxId.serialize() 
 * const deser = CTxId.deserialize(ser)
 * ser === deser.serialize() // true
 * ```
 */
export class CTxId extends ManagedObj {
  constructor(obj: any) {
    super(obj)
  }

  override value(): any {
    return castToUint8_tPtr(this.obj)
  }

  override serialize(): string {
    const buf = this.value()
    return toHex(buf, CTX_ID_SIZE)
  }

  /** Deserializes a `CTxId` from its hexadecimal representation.
   *
   * * @param hex - The hexadecimal string to deserialize.
   * * @returns A new `CTxId` instance.
   */
  static deserialize(
    this: new (obj: any) => CTxId,
    hex: string
  ): CTxId {
    if (hex.length % 2 !== 0) {
      hex = `0${hex}`
    }
    if (hex.length !== CTX_ID_SIZE * 2) {
      throw new Error(`Invalir TxId hex length. Expected ${CTX_ID_SIZE * 2}, but got ${hex.length}.`)
    }

    const obj = hexToMallocedBuf(hex) 
    return new CTxId(obj)
  }
}

